tracecat/auth/saml.py

```diff
@@ -0,0 +1,289 @@
+        if attr_statement is None:
+            logger.error("SAML response failed: AttributeStatement not found")
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid SAML response"
+            ) from None
+
+        # Process all attributes
+        attributes = {}
+        for attr_elem in attr_statement.findall("saml2:Attribute", self.NAMESPACES):
+            saml_attr = self._extract_attribute(attr_elem)
+            attributes[saml_attr.name] = asdict(saml_attr)
+
+        return attributes
+
+
+@contextmanager
+def generate_saml_metadata_file():
+    """Generate a temporary SAML metadata file."""
+
+    if not SAML_IDP_ENTITY_ID:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="SAML SSO entity ID has not been configured.",
+        )
+
+    if not SAML_IDP_REDIRECT_URL:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="SAML SSO redirect URL has not been configured.",
+        )
+
+    if not SAML_IDP_CERTIFICATE:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="SAML SSO certificate has not been configured.",
+        )
+
+    # Create the root element
+    root = ET.Element(
+        "EntityDescriptor",
+        {
+            "xmlns": "urn:oasis:names:tc:SAML:2.0:metadata",
+            "xmlns:ds": "http://www.w3.org/2000/09/xmldsig#",
+            "entityID": SAML_IDP_ENTITY_ID,
+        },
+    )
+
+    # Create IDPSSODescriptor element
+    idp_sso_descriptor = ET.SubElement(
+        root,
+        "IDPSSODescriptor",
+        {"protocolSupportEnumeration": "urn:oasis:names:tc:SAML:2.0:protocol"},
+    )
+
+    # Add KeyDescriptor
+    key_descriptor = ET.SubElement(idp_sso_descriptor, "KeyDescriptor", use="signing")
+    key_info = ET.SubElement(key_descriptor, "ds:KeyInfo")
+    x509_data = ET.SubElement(key_info, "ds:X509Data")
+    x509_certificate = ET.SubElement(x509_data, "ds:X509Certificate")
+    x509_certificate.text = SAML_IDP_CERTIFICATE
+
+    # Add NameIDFormat
+    name_id_format = ET.SubElement(idp_sso_descriptor, "NameIDFormat")
+    name_id_format.text = "urn:oasis:names:tc:SAML:2.0:nameid-format:persistent"
+
+    # Add SingleSignOnService
+    ET.SubElement(
+        idp_sso_descriptor,
+        "SingleSignOnService",
+        {"Binding": BINDING_HTTP_REDIRECT, "Location": SAML_IDP_REDIRECT_URL},
+    )
+
+    # Create a temporary file
+    with tempfile.NamedTemporaryFile(
+        mode="w+", suffix=".xml", delete=False
+    ) as tmp_file:
+        # Write the XML to the temporary file
+        tree = ET.ElementTree(root)
+        tree.write(tmp_file, encoding="unicode", xml_declaration=True)
+        tmp_file.flush()
+        tmp_file_path = tmp_file.name
+
+    try:
+        yield tmp_file_path
+    finally:
+        # Clean up the temporary file
+        os.unlink(tmp_file_path)
```

Comment:
nit: you could probably just do this, to the same effect
```suggestion
    # Create a temporary file
    with tempfile.NamedTemporaryFile(mode="w+", suffix=".xml") as tmp_file:
        # Write the XML to the temporary file
        tree = ET.ElementTree(root)
        tree.write(tmp_file, encoding="unicode", xml_declaration=True)
        tmp_file.flush()
        tmp_file_path = tmp_file.name
        yield tmp_file_path
```